Aks autoscale

jlpedrosa opened the pull request:
Hi

This PR adds auto scale and availability zones capabilities to AKS resource and data providers.

It bumps the API version for AKS, there are other PR pending regarding that matter so I expect a conflict depends on what gets merged first. 
[https://github.com/terraform-providers/terraform-provider-azurerm/pull/3262](url)

This depends on the following features:
az feature register --namespace Microsoft.ContainerService -n AvailabilityZonePreview
az feature register --namespace Microsoft.ContainerService --name VMSSPreview 
az provider register -n Microsoft.ContainerService

katbyte commented:
Duplicate of #3174

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -109,6 +109,34 @@ func resourceArmKubernetesCluster() *schema.Resource {
+						"max_count": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
```
Could we get some validation here? Not sure of the exact values but:
```suggestion
							ValidateFunc: validation.IntBetween(1, 100),
						},
```

jlpedrosa replied:
Done

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -109,6 +109,34 @@ func resourceArmKubernetesCluster() *schema.Resource {
+						"min_count": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
```
Could we get some validation here? Not sure of the exact values but:
```suggestion
							ValidateFunc: validation.IntBetween(1, 100),
						},
```

jlpedrosa replied:
Done

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -109,6 +109,34 @@ func resourceArmKubernetesCluster() *schema.Resource {
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
```
Needs validation but i think master has this property now

jlpedrosa replied:
YEs master had it now.

katbyte commented on azurerm/resource_arm_kubernetes_cluster_test.go:
```diff
@@ -548,6 +548,65 @@ func TestAccAzureRMKubernetesCluster_advancedNetworkingAzureNPMPolicyComplete(t
+
+}
+
+func TestAccAzureRMKubernetesCluster_autoScalingWithAvailabilityZones(t *testing.T) {
```
Could we get another test that updates these properties?

jlpedrosa replied:
as I mentioned in https://github.com/terraform-providers/terraform-provider-azurerm/pull/317 the moment we change the type of agent pool to VMScaleSets (which is required for autoscaling, but it can be enabled withoutautoscaling). Then any put request to the api will fail (which means no updates will work with that type of agent pool).

katbyte commented on azurerm/resource_arm_kubernetes_cluster_test.go:
```diff
@@ -1358,3 +1417,64 @@ resource "azurerm_kubernetes_cluster" "test" {
+
+  agent_pool_profile {
+    name    = "default"    
+		vm_size = "Standard_DS2_v2"
```
Could we run this through `terraform fmt` to fix the tabs/indentation

jlpedrosa replied:
Done

katbyte commented on azurerm/resource_arm_kubernetes_cluster_test.go:
```diff
@@ -1358,3 +1417,64 @@ resource "azurerm_kubernetes_cluster" "test" {
+  dns_prefix          = "acctestaks%d"
+
+  agent_pool_profile {
+    name    = "default"    
```
Same here

jlpedrosa replied:
Done

jlpedrosa commented:
Hi @katbyte 

Your guess was correct regarding the limits, in the docs the thershold is 100 [https://docs.microsoft.com/en-us/azure/aks/quotas-skus-regions](aks limits)

I've rebased the PR (there were some conflicts).
Added the validations
Added the documentation
terraform fmt the test code (taking it to a file outside .go code and fmt then copying it back).

regarding: "test that updates the properties", as I mentioned in https://github.com/terraform-providers/terraform-provider-azurerm/pull/317 the moment we change the type of agent pool to VMScaleSets (which is required for autoscaling, but it can be enabled withoutautoscaling). Then any put request to the api will fail (which means no updates will work with that type of agent pool).

katbyte commented:
FYI, i have a tool that will run `terraform fmt` on blocks of tf code in the tests files: https://github.com/katbyte/terrafmt

katbyte commented on azurerm/data_source_kubernetes_cluster.go:
```diff
@@ -564,6 +587,26 @@ func flattenKubernetesClusterDataSourceAgentPoolProfiles(input *[]containerservi
+			agentPoolProfile["max_count"] = int(*profile.MaxCount)
+		}
+
+		if profile.Type != "" {
```
This is already set above on line 582

jlpedrosa replied:
Correct. Merge issue, re added in one of the rebases.

jlpedrosa replied:
this was fixed days I go.

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -934,6 +959,46 @@ func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) []containe
+			profile.MinCount = utils.Int32(minCount)
+		}
+
+		if vmSetType := config["type"].(string); vmSetType != "" {
```
THis is already set on line 971

jlpedrosa replied:
Fixed, was a rebase issue

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -934,6 +959,46 @@ func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) []containe
+			profile.AvailabilityZones = availavilityZones
+		}
+
+		if maxCount := int32(config["max_count"].(int)); maxCount > 0 {
```
Isn't this already set on line 963?

jlpedrosa replied:
Rebase issue

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -934,6 +959,46 @@ func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) []containe
+			profile.MaxCount = utils.Int32(maxCount)
+		}
+
+		if minCount := int32(config["min_count"].(int)); minCount > 0 {
```
Isn't this already set on line 967?

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -934,6 +959,46 @@ func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) []containe
+			profile.Type = containerservice.AgentPoolType(vmSetType)
+		}
+
+		if enableAutoScalingItf := config["enable_auto_scaling"]; enableAutoScalingItf != nil {
```
ditto

jlpedrosa replied:
Rebase issue

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -934,6 +959,46 @@ func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) []containe
+			profile.EnableAutoScaling = utils.Bool(enableAutoScalingItf.(bool))
+		}
+
+		if availavilityZones := utils.ExpandStringSlice(config["availability_zones"].([]interface{})); len(*availavilityZones) > 0 {
```
ditto

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -957,6 +1022,26 @@ func flattenKubernetesClusterAgentPoolProfiles(profiles *[]containerservice.Mana
+			agentPoolProfile["max_count"] = int(*profile.MaxCount)
+		}
+
+		if profile.Type != "" {
```
I believe this has already been set to?

jlpedrosa replied:
Rebase issue

katbyte commented on azurerm/resource_arm_kubernetes_cluster_test.go:
```diff
@@ -1480,6 +1480,8 @@ resource "azurerm_kubernetes_cluster" "test" {
     count   = "1"
     vm_size = "Standard_DS2_v2"
+    name    = "default"    
+		vm_size = "Standard_DS2_v2"
```
```suggestion
    vm_size = "Standard_DS2_v2"
```

katbyte commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
 
 * `max_pods` - The maximum number of pods that can run on each agent.
 
+* `availability_zones` - (Optional)  Availability zones for nodes. Must use VirtualMachineScaleSets for `type` of the `agent_pool_profile`.
```
Could we word this like other data source:
```suggestion
* `availability_zones` - The availability zones used for the nodes. 
```

katbyte commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
 
+* `availability_zones` - (Optional)  Availability zones for nodes. Must use VirtualMachineScaleSets for `type` of the `agent_pool_profile`.
+
+* `enable_auto_scaling` - (Optional) Whether to enable auto-scaler [Autoscaling details](https://docs.microsoft.com/en-us/azure/aks/cluster-autoscaler) 
```
```suggestion
* `enable_auto_scaling` - If the auto-scaler is enabled. 
```

katbyte commented on website/docs/r/kubernetes_cluster.html.markdown:
```diff
@@ -120,14 +120,24 @@ A `agent_pool_profile` block supports the following:
 
 * `max_pods` - (Optional) The maximum number of pods that can run on each agent.
+ 
+* `availability_zones` - (Optional)  Availability zones for nodes. Must use VirtualMachineScaleSets for `type` of the `agent_pool_profile`.
```
```suggestion
* `availability_zones` - (Optional)  Availability zones for nodes. The property `type` of the `agent_pool_profile` must be set to `VirtualMachineScaleSets` to use this property.
```

funnybirdman commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
+
+~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
+
+* `min_coumt` - (Optional) Minimum number of nodes for auto-scaling 
```
```suggestion
* `min_count` - (Optional) Minimum number of nodes for auto-scaling 
```

jlpedrosa replied:
Thanks!

jlpedrosa commented:
@tombuildsstuff @jlpedrosa 

I ran the tests against my subscription, everything seemed fine. Do you know when it could me merged?

tombuildsstuff commented on website/docs/r/kubernetes_cluster.html.markdown:
```diff
@@ -120,14 +120,24 @@ A `agent_pool_profile` block supports the following:
+
+~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
+
+* `min_coumt` - (Optional) Minimum number of nodes for auto-scaling 
```
```suggestion
* `min_count` - (Optional) Minimum number of nodes for auto-scaling 
```

jlpedrosa replied:
Fixed in the commit.

tombuildsstuff commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
+
+* `enable_auto_scaling` - If the auto-scaler is enabled. 
+
+~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
```
since this is a Data Source I think we can remove this `note` here

```suggestion
```

tombuildsstuff commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
+
+~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
+
+* `min_count` - (Optional) Minimum number of nodes for auto-scaling 
```
since this is a Data Source this is read-only - so could we remove this:

```suggestion
* `min_count` - Minimum number of nodes for auto-scaling 
```

jlpedrosa replied:
Fixed in the commit.

tombuildsstuff commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
+
+* `min_count` - (Optional) Minimum number of nodes for auto-scaling 
+
+* `max_count` - (Optional) Maximum number of nodes for auto-scaling
```
since this is a Data Source this is read-only - so could we remove this:

```suggestion
* `max_count` - Maximum number of nodes for auto-scaling
```

jlpedrosa replied:
Fixed.

tombuildsstuff commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -957,6 +999,22 @@ func flattenKubernetesClusterAgentPoolProfiles(profiles *[]containerservice.Mana
+
+		if profile.AvailabilityZones != nil && len(*profile.AvailabilityZones) > 0 {
+			agentPoolProfile["availability_zones"] = utils.FlattenStringSlice(profile.AvailabilityZones)
+		}
```
can we remove this if statement around this flatten? the `utils.FlattenStringSlice` should handle this being nil (and we should always set this complex type if there's no elements, since that'll then show a diff)

jlpedrosa replied:
@tombuildsstuff:
In this case,  I am protecting us against some behaviours of the API where an empty array is not the same as not sending an array.

katbyte replied:
But here we are setting the terraform side. Any protection on the API side should be where we set that property on the SDK (only set it if tf data is !nil and len > 0)

jlpedrosa replied:
I removed it here and just left it on the expand. Is that OK?

tombuildsstuff commented on azurerm/data_source_kubernetes_cluster.go:
```diff
@@ -564,6 +587,22 @@ func flattenKubernetesClusterDataSourceAgentPoolProfiles(input *[]containerservi
+
+		if profile.AvailabilityZones != nil {
+			agentPoolProfile["availability_zones"] = utils.FlattenStringSlice(profile.AvailabilityZones)
+		}
```
could we remove this `if` statement here? the Flatten function handles this being nil and returning an empty list

jlpedrosa replied:
This is the other side of not sending empty lists to the API, I don't want to store the empty entry en the map so later I don't actually send it when invoking the api

katbyte replied:
This shouldn't matter as its a datasource (and in the expand function you check for len >0 anyways)

jlpedrosa replied:
Agreed, as stated in the other comment all fields were following this pattern. Removed already

tombuildsstuff commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -933,6 +958,23 @@ func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) []containe
+
+		if enableAutoScalingItf := config["enable_auto_scaling"]; enableAutoScalingItf != nil {
+			profile.EnableAutoScaling = utils.Bool(enableAutoScalingItf.(bool))
+		}
```
there's a couple of things here:

1. if auto-scaling's enabled and we don't have a min/max set - should we raise an error here? This'd require updating the function to also support returning/raising errors:

```
func expandKubernetesClusterAgentPoolProfiles(d *schema.ResourceData) ([]containerservice.ManagedClusterAgentPoolProfile, error) {
```

(for which we can return `nil` by default for the error)

2. if auto-scaling is enabled; should we conditionally not be setting the 'count' field? I'm wondering if that will force a node-pool resizing during an update?

jlpedrosa replied:
Hi @tombuildsstuff 
Fully agree with 1) I'll add it to the patch.
about the 2) is more complicated: if you say { min:1, max:3, count:2 }, The cluster is created with 2 nodes. If you don't send the count you get:
```
{
    "code": "InvalidParameter",
    "message": "The value of parameter agentPoolProfile.count is invalid. Please see https://aka.ms/aks-naming-rules for more details.",
    "target": "agentPoolProfile.count"
}
```

The issue comes with the updates, despite that is the same HTTP methos and URL the api has some  really different behaviour depending the type of agent pool (not related to autoscaling).

Any update to the AKS  cluster if the node pool is  `VirtualMachineScaleSets` will throw:
```
{
    "code": "OperationNotAllowed",
    "message": "Node pool update via managed cluster not allowed. Use per nodepool operations."
}
```

As soon as that is clarified with the AKS team, then we can go left and handle the exception in terraform, or they'll make the API to behave the same regardless the type of agentpool.

tombuildsstuff commented on README.md:
````diff
@@ -94,7 +94,7 @@ $ make test
 
 ```
-make testacc TESTARGS='-run=TestAccAzureRMResourceGroup'
+make testacc TESTARGS=-run=TestAccAzureRMResourceGroup
````
can we revert this? whilst the quotes don't matter on macOS and Linux - they're mostly here for Windows support (unless you're using Windows and this is working for you, in which case ignore this 🙃)

jlpedrosa replied:
I was giving me issues in Mac, that is why I thought it was a typo. Fixed in a new push.

jlpedrosa commented on azurerm/data_source_kubernetes_cluster.go:
```diff
@@ -564,6 +587,22 @@ func flattenKubernetesClusterDataSourceAgentPoolProfiles(input *[]containerservi
+
+		if profile.AvailabilityZones != nil {
+			agentPoolProfile["availability_zones"] = utils.FlattenStringSlice(profile.AvailabilityZones)
+		}
```
@tombuildsstuff:
I was trying to avoid creating emepy entries in the map (agentPoolProfile). If profile.AvailabilityZones is nil, then we'll get an empty entry on the map. [Snippet for better explanation](https://play.golang.org/p/6Trfl-loUEf) 
Let me know what you think and I'll do it as instructed.

katbyte replied:
As this is on the TF side what is the issue if we set this to an empty array?

jlpedrosa replied:
In the data source none. The rest of the code in that provider is following that pattern. Not sure if nicer in the state files (if it actually goes there). So I'll remove it, if you want it back just let me know.

jlpedrosa commented:
Hi @tombuildsstuff 

I've rebased it, but your comments were addressed few days ago, I don't know if you could have a look.

katbyte commented on website/docs/r/kubernetes_cluster.html.markdown:
```diff
@@ -120,14 +120,24 @@ A `agent_pool_profile` block supports the following:
 * `type` - (Optional) Type of the Agent Pool. Possible values are `AvailabilitySet` and `VirtualMachineScaleSets`. Changing this forces a new resource to be created. Defaults to `AvailabilitySet`.
 
-~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. To use it, enable feature `VMSSPreview` for `namespace Microsoft.ContainerService`. For an example of how to enable a Preview feature, please visit [Register scale set feature provider](https://docs.microsoft.com/en-us/azure/aks/cluster-autoscaler#register-scale-set-feature-provide ...
+~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
```
Could we not change this line? we recently updated this wording in master

jlpedrosa replied:
Fixed.

katbyte commented on website/docs/r/kubernetes_cluster.html.markdown:
```diff
@@ -120,14 +120,24 @@ A `agent_pool_profile` block supports the following:
+
+* `enable_auto_scaling` - (Optional) Whether to enable auto-scaler [Autoscaling details](https://docs.microsoft.com/en-us/azure/aks/cluster-autoscaler) 
+
+~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
```
This line is spurious as its duplicated below?

jlpedrosa replied:
Hi
if you mean 
```
~>  **Note:** Support for the `type` of `VirtualMachineScaleSets` is currently in Public Preview on an opt-in basis. You can enable this feature using the Azure CLI by running `az feature register --name VMSSPreview --namespace Microsoft.ContainerService`
```

No, it's not duplicated, it is a requirement for autoscale sets, it only works with VirtualMachineScaleSets. I've added more comments in the doc. Let me know if now is clear or how would you preffer it. I could add also checks in the code.?

katbyte commented on azurerm/resource_arm_kubernetes_cluster.go:
```diff
@@ -563,7 +588,12 @@ func resourceArmKubernetesClusterCreateUpdate(d *schema.ResourceData, meta inter
 	linuxProfile := expandKubernetesClusterLinuxProfile(d)
-	agentProfiles := expandKubernetesClusterAgentPoolProfiles(d)
+	agentProfiles, err := expandKubernetesClusterAgentPoolProfiles(d)
+
```
```suggestion
```

katbyte commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
+
+* `min_count` - Minimum number of nodes for auto-scaling 
+
+* `min_count` - (Optional) Minimum number of nodes for auto-scaling 
```
As this is a data source could we remove:
```suggestion
* `min_count` - Minimum number of nodes for auto-scaling 
```

katbyte commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -86,6 +86,16 @@ A `agent_pool_profile` block exports the following:
+
+* `min_count` - (Optional) Minimum number of nodes for auto-scaling 
+
+* `max_count` - (Optional) Maximum number of nodes for auto-scaling
```
As this is a data source could we remove:
```suggestion
* `max_count` - Maximum number of nodes for auto-scaling
```

katbyte commented on website/docs/r/kubernetes_cluster.html.markdown:
```diff
@@ -120,14 +120,24 @@ A `agent_pool_profile` block supports the following:
+ 
+* `availability_zones` - (Optional)  Availability zones for nodes. The property `type` of the `agent_pool_profile` must be set to `VirtualMachineScaleSets` to use this property.
+
+* `enable_auto_scaling` - (Optional) Whether to enable auto-scaler [Autoscaling details](https://docs.microsoft.com/en-us/azure/aks/cluster-autoscaler) 
```
This might be better as:
```suggestion
* `enable_auto_scaling` - (Optional) Whether to enable [auto-scaler](https://docs.microsoft.com/en-us/azure/aks/cluster-autoscaler).
```

jlpedrosa commented:
Hi @katbyte 

I've fixed most of the comments, and final questions to few. Please your feedback, I've rebased and squashed it as it was giving merge errors. Sorry for the delay.

TVH7 commented:
When will the new release be pushed? Waiting for this..

TVH7 commented:
@katbyte I've seen that the milestone deadline was yesterday however no new release has been made. Any updates?

PatricioDanos commented:
I'm also waiting for this release. Is there any news when it's going to be GA?

TVH7 commented:
> I'm also waiting for this release. Is there any news when it's going to be GA?

So according to my latest info they have a milestone open for the next release. This is already a week overdue (although it says 2 days overdue, but that is because they keep moving the due date.) 

However, we are now at 97% complete so I'm expecting @katbyte to release today or tomorrow (24th/25th of July) 

A confirmation by @katbyte would be appreciated though.

PatricioDanos commented:
> > I'm also waiting for this release. Is there any news when it's going to be GA?
> 
> So according to my latest info they have a milestone open for the next release. This is already a week overdue (although it says 2 days overdue, but that is because they keep moving the due date.)
> 
> However, we are now at 97% complete so I'm expecting @katbyte to release today or tomorrow (24th/25th of July)
> 
> A confirmation by @katbyte would be appreciated though.
> 
> Link to the milestone: https://github.com/terraform-providers/terraform-provider-azurerm/milestone/48

We're now at 100% so it should be released asap.
